src/scitokens_internal.cpp

```diff
@@ -792,11 +794,15 @@ Validator::get_public_key_pem(const std::string &issuer, const std::string &kid,
 
     if (get_public_keys_from_db(issuer, now, result->m_keys,
                                 result->m_next_update)) {
-        if (now > result->m_next_update) {
+        std::unique_lock<std::mutex> lock(key_refresh_mutex, std::defer_lock);
+        // If refresh is due *and* the key refresh mutex is free, try to update
+        if (now > result->m_next_update && lock.try_lock()) {
             try {
                 result->m_ignore_error = true;
                 result = get_public_keys_from_web(
                     issuer, internal::SimpleCurlGet::default_timeout);
+                // Hold refresh mutex in the new result
+                result->m_refresh_lock = std::move(lock);
```

Comment:
Why are we storing this in the result?  As far as I can tell, it's never used.

Reply:
The lock is moved to the result object, so the mutex is held throughout the lifetime of the HTTP request. When the result is stored and destroyed, the lock is released.

Reply:
Ah, ok, making use of the implicit destructor.  gotcha.